hydra-node/test/Hydra/JSONSchema.hs

```diff
@@ -0,0 +1,150 @@
+  (ToJSON a, Arbitrary a, Show a) =>
+  FilePath ->
+  FilePath ->
+  Property
+prop_validateToJSON specFile inputFile =
+  forAllShrink (vectorOf 100 arbitrary) shrink $ \(a :: [a]) ->
+    monadicIO $
+      do
+        run ensureSystemRequirements
+        let obj = Aeson.encode a
+        (exitCode, _out, err) <- run $ do
+          writeFileLBS inputFile obj
+          readProcessWithExitCode "jsonschema" ["-i", inputFile, specFile] mempty
+        monitor $ counterexample err
+        monitor $ counterexample (decodeUtf8 obj)
+        assert (exitCode == ExitSuccess)
+
+-- | Check specification is complete wr.t. to generated data
+-- This second sub-property ensures that any key found in the
+-- specification corresponds to a constructor in the corresponding
+-- data-type. This in order the document in sync and make sure we don't
+-- left behind constructors which no longer exists.
+--
+-- The second argument is a lens-like function that says which part of the
+-- specification to select to check completeness of the specification w.r.t.
+-- constructors for the datatype, for example:
+--
+-- @@
+-- key "properties" . key "message" . key "oneOf" . _Array . to toList
+-- @@
+--
+-- which selects the list of elements under @properties > message > oneOf@ path
+-- in the specification file.
+prop_specIsComplete ::
+  forall a.
+  (Arbitrary a, Show a) =>
+  FilePath ->
+  SpecificationSelector Yaml.Value ->
+  Property
+prop_specIsComplete specFile typeSpecificationSelector =
+  forAllBlind (vectorOf 1000 arbitrary) $ \(a :: [a]) ->
+    monadicIO $ do
+      specs <- run $ Aeson.decodeFileStrict specFile
+      let knownKeys = classify specs a
+      let unknownConstructors = Map.keys $ Map.filter (== 0) knownKeys
+
+      when (null knownKeys) $ do
+        monitor $ counterexample "No keys found in given specification fragment"
+        assert False
+
+      unless (null unknownConstructors) $ do
+        let commaSeparated = intercalate ", " (toString <$> unknownConstructors)
+        monitor $ counterexample $ "Unimplemented constructors present in specification: " <> commaSeparated
+        monitor $ counterexample $ show a
+        assert False
+ where
+  -- Like Generics, if you squint hard-enough.
+  strawmanGetConstr :: a -> Text
+  strawmanGetConstr = toText . Prelude.head . words . show
+
+  classify :: Maybe Aeson.Value -> [a] -> Map Text Integer
+  classify (Just specs) =
+    let knownKeys =
+          case specs ^? typeSpecificationSelector of
+            Just es ->
+              let ks = mapMaybe (\(e :: Aeson.Value) -> e ^? key "title" . _String) es
+               in Map.fromList $ zip ks (repeat @Integer 0)
+            _ ->
+              mempty
+
+        countMatch (strawmanGetConstr -> tag) =
+          Map.alter (Just . maybe 1 (+ 1)) tag
+     in foldr countMatch knownKeys
+  classify _ =
+    error $ "Invalid specification file. Does not decode to an object: " <> show specFile
+
+type SpecificationSelector t =
```

Comment:
Why is `t` is not used in the definition of the type alias?

Also, I don't understand this. Is this is a lens or a traversal?

Also, I would expect the helper function `apiSpecificationSelector` and some documentation how/when to use it here.. if you already go through the hassle of accepting the not only the key to `prop_specIsComplete`